src/thor/bidirectional_astar.cc

```diff
@@ -916,10 +918,167 @@ void BidirectionalAStar::SetDestination(GraphReader& graphreader, const valhalla
   }
 }
 
+/**
+ * Recover shortcut edge and recost it
+ * @param  graphreader  Graph tile reader.
+ * @param  shortcut_id  Edge id of the shortcut to be recosted
+ * @param  costing      Costing function
+ * @param  transition_cost  Transition cost for the shortcut edge
+ * @param  previous_cost    Total cost from the previous edge
+ * TODO: add time information
+ */
+std::vector<PathInfo> recost_shortcut_forward(GraphReader& graphreader,
+                                              const GraphId& shortcut_id,
+                                              const DynamicCost& costing,
+                                              Cost transition_cost,
+                                              Cost previous_cost) {
+  auto edges = graphreader.RecoverShortcut(shortcut_id);
```

Comment:
as soon as this fails should we log and return?

Reply:
we already log errors in `RecoverShortcut` method, so I think we shouldn't log here. but, it might be reasonable not to continue and treat recovering fail as recosting fail - in this case no necessary to call `recost_forward` function becasue we can use known shortcut cost.